interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+                if let Some(val) = hm.get(k) {
+                    return Some(val);
+                }
+            }
+            //then check if it's anywhere in the other list
+            return None;
+        }
+    }
+
+    /// ```text
+    /// Sets a new binding of [k] to [v] in the highest scope.
+    /// Guarantees that all following calls to [get(&k)] will return [&v],
+    /// granted no new binding is set with [set]
+    /// ```
+    /// # Example
+    /// Using one scope:
+    /// ```rust
+    /// use interp::stk_env::Smoosher;
+    /// let mut smoosher = Smoosher::new();
+    /// smoosher.set("hi!", 1);
+    /// assert_eq!(*smoosher.get(&"hi!").unwrap(), 1);
+    /// smoosher.set("hi!", 2);
+    /// assert_eq!(*smoosher.get(&"hi!").unwrap(), 2);
+    /// ```
+    /// More than one scope:
+    /// ```rust
+    /// use interp::stk_env::Smoosher;
+    /// let mut smoosher = Smoosher::new();
+    /// smoosher.set("hi!", 1);
+    /// assert_eq!(*smoosher.get(&"hi!").unwrap(), 1);
+    /// smoosher.new_scope(); //scopes themselves do not affect gets, only new sets do.
+    /// assert_eq!(*smoosher.get(&"hi!").unwrap(), 1);
+    /// smoosher.set("hi!", 2);
+    /// assert_eq!(*smoosher.get(&"hi!").unwrap(), 2);
+    /// ```
+    pub fn set(&mut self, k: K, v: V) {
+        self.head.insert(k, v);
+    }
+
+    /// ```text
+    /// Returns a new Smoosher and mutates [self]. The new Smoosher has a new scope
+    /// as [head] and all of (pre-mutation) [self] as [tail]. [Self] has a fresh scope pushed onto
+    /// it. Invariant this method enforces: you cannot mutate a scope that has children
+    /// forks (you can only mutate the fresh scope applied atop it)
+    /// ```
+    /// # Examples
+    /// ## Pictorial Example
+    /// ```text
+    /// [A]
+    /// ```
+    /// let B = A.fork();
+    /// ```text
+    /// [B] [A'] <- Both B and A' point to empty nodes spawned off A
+    ///  |   |
+    ///   \ /
+    ///    |
+    ///   [A]
+    /// ```
+    /// ## Code Example
+    /// ```rust
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// let mut b = a.fork();
+    /// a.set("hey", 2);
+    /// b.set("hey", 3);
+    /// assert_eq!(*a.get(&"hi!").unwrap(), *b.get(&"hi!").unwrap());
+    /// assert_eq!(*a.get(&"hey").unwrap(), 2);
+    /// assert_eq!(*b.get(&"hey").unwrap(), 3);
+    /// ```
+    pub fn fork(&mut self) -> Self {
+        //first save self's head, and replace it with a clean HM
+        let old_head = mem::replace(&mut self.head, HashMap::new()); //can replace with mem::take()
+                                                                     //create tail to both Self and the new Smoosher
+        let new_tail = self.tail.push(old_head);
+        //update Self and create new Smoosher
+        self.tail = new_tail.clone(); //will this die after the end of [fork]? no
+        Smoosher {
+            head: HashMap::new(),
+            tail: new_tail,
+        }
```

Comment:
You could create a constructor for this specifically, as a convenience

Reply:
This is only used here, I think